sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MobileDatabaseEncryptionKeysCheck.kt

```diff
@@ -0,0 +1,94 @@
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.DescriptorToSourceUtils
+import org.jetbrains.kotlin.resolve.calls.callUtil.getResolvedCall
+import org.jetbrains.kotlin.resolve.calls.model.ExpressionValueArgument
+import org.jetbrains.kotlin.resolve.calls.model.ResolvedCall
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.CallAbstractCheck
+import org.sonarsource.kotlin.api.ConstructorMatcher
+import org.sonarsource.kotlin.api.FunMatcher
+import org.sonarsource.kotlin.api.SecondaryLocation
+import org.sonarsource.kotlin.api.predictRuntimeValueExpression
+import org.sonarsource.kotlin.converter.KotlinTextRanges.textRange
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+private val CREATE_CHAR_BYTE_ARRAY = FunMatcher(qualifier = "kotlin") { withNames("byteArrayOf", "charArrayOf") }
+private const val SQLITE = "net.sqlcipher.database.SQLiteDatabase"
+private const val ENCRYPTION_KEY = "encryptionKey"
+private const val CHANGE_PASSWORD = "changePassword"
+
+@Rule(key = "S6301")
+class MobileDatabaseEncryptionKeysCheck : CallAbstractCheck() {
+    override fun functionsToVisit() = listOf(
+        ConstructorMatcher(typeName = SQLITE),
+        FunMatcher(qualifier = SQLITE) {
+            withNames(CHANGE_PASSWORD, "openDatabase", "openOrCreateDatabase", "create")
+        },
+        FunMatcher(qualifier = "io.realm.RealmConfiguration.Builder", name = ENCRYPTION_KEY)
+    )
+
+    override fun visitFunctionCall(
+        callExpression: KtCallExpression,
+        resolvedCall: ResolvedCall<*>,
+        kotlinFileContext: KotlinFileContext
+    ) {
+        val bindingContext = kotlinFileContext.bindingContext
+        val functionName = resolvedCall.resultingDescriptor.name.asString()
+
+        val valueArgumentsList = resolvedCall.valueArgumentsByIndex
+        if (valueArgumentsList == null || (valueArgumentsList.size < 2 && functionName != ENCRYPTION_KEY)) return
+
+        val arg = if (functionName in setOf(ENCRYPTION_KEY, CHANGE_PASSWORD)) {
+            valueArgumentsList[0]
+        } else valueArgumentsList[1]
+        val argExpr = (arg as? ExpressionValueArgument)?.valueArgument?.getArgumentExpression() ?: return
+
+        val secondaries = mutableListOf<PsiElement>()
+        val argValueExpr = argExpr.predictRuntimeValueExpression(bindingContext, secondaries)
+        if (argValueExpr.isHardCoded(bindingContext, secondaries)) {
+            val parameter = if (functionName == ENCRYPTION_KEY) ENCRYPTION_KEY else "password"
+
+            kotlinFileContext.reportIssue(
+                callExpression.calleeExpression!!,
+                """The "$parameter" parameter should not be hardcoded.""",
+                secondaries.map { SecondaryLocation(kotlinFileContext.textRange(it)) },
+            )
+        }
+    }
+}
+
+private fun KtExpression.isHardCoded(bindingContext: BindingContext, secondaries: MutableList<PsiElement>) : Boolean =
```

Comment:
Do expressions in parentheses work here? So e.g. `val foo = ("hardcoded")`? What about `val foo = (((((((("hardcoded"))))))))`?